Add a Recordable interface, implementation, and tests for logs

xukaren opened the pull request:
This PR implements a recordable for logs, as part of the [logging prototype](https://github.com/open-telemetry/opentelemetry-cpp/issues/337), and also resolving issue #412. 

This PR:

* Adds a Recordable interface for logs
* Add a default `LogRecord` Recordable  implementation, with unit tests 
* Changes the methods and signatures to take in a `Recordable` instead of the previous `LogRecord` struct, across the processor and exporter files
* Refactors all affected files (e.g. tests for the logging API, SDK, and processors) 

API/SDK changes: 
* Gets the TraceId, SpanId, TraceFlags, and current Timestamp in the SDK and injects into the LogRecord if it is not user-specified. 
* Removed the `Severity::kDefault` enum value (originally set as kInfo), and added a `Severity::kInvalid`= 0. (Setting the default Severity to kInfo has been moved to the SDK instead of the API)

Other minor changes: 
* Removes the `LogRecord` struct from the API 
* Stylistic change: Renames the `log()` methods to `Log()`, and adds `this->` before calling Log() methods
 
This PR introduces changes that blocks a few other open PRs: #422, #430, #434, and #435.

cc @MarkSeufert @alolita

maxgolov commented on api/include/opentelemetry/logs/logger.h:
```diff
@@ -43,81 +60,105 @@ class Logger
   {
-    // TODO: set default timestamp later (not in API)
-    log(severity, message, core::SystemTimestamp(std::chrono::system_clock::now()));
+    return this->Log(timestamp, severity, name, body,
```
Is there a practical reason why you need `this->` here ?

xukaren replied:
Not really, but I thought it would be better to follow the convention of [StartSpan()](https://github.com/open-telemetry/opentelemetry-cpp/blob/master/api/include/opentelemetry/trace/tracer.h#L52)

maxgolov commented on api/include/opentelemetry/logs/logger.h:
```diff
@@ -43,81 +60,105 @@ class Logger
+           Severity severity,
+           nostd::string_view name,
+           nostd::string_view body,
+           const T &resources,
```
Should this be `resource` ?
```suggestion
           const T &resource,
```

maxgolov commented on api/include/opentelemetry/logs/logger.h:
```diff
@@ -43,81 +60,105 @@ class Logger
   {
-    // Set severity to the default then call log(Severity, String message) method
-    log(Severity::kDefault, message);
+    Log(timestamp, severity, name, body, common::KeyValueIterableView<T>(resources),
```
```suggestion
    Log(timestamp, severity, name, body, common::KeyValueIterableView<T>(resource),
```

maxgolov commented on api/include/opentelemetry/logs/logger.h:
```diff
@@ -43,81 +60,105 @@ class Logger
+           Severity severity,
+           nostd::string_view name,
+           nostd::string_view body,
+           std::initializer_list<std::pair<nostd::string_view, common::AttributeValue>> resource,
```
Basically either `resources` here or `resource` above for the sake of consistency.

xukaren replied:
Typo, my mistake - it should be `resource` to match the [Log Data Model field](https://github.com/open-telemetry/opentelemetry-specification/blob/master/specification/logs/data-model.md#log-and-event-record-definition)

maxgolov commented on api/include/opentelemetry/logs/logger.h:
```diff
@@ -43,81 +60,105 @@ class Logger
+  inline void Log(Severity severity, nostd::string_view message) noexcept
   {
-    log(severity, name, common::KeyValueIterableView<T>(attributes));
+    this->Log(severity, message, default_timestamp_);
```
I do not have a strong opinion on this. But ideally the `default_timestamp_` should be something like `std::chrono::system_clock::now()`? Maybe we can make an assumption that duration of time `inline` call to `Log` and to the moment when we set it to that actual value (in Exporter?) is rather miniscule.. Basically we log `current name` by default. Maybe an implementation of overridable getter that allows to populate this in concrete SDK logger implementation.

lalitb replied:
+1

xukaren replied:
True, updated this in the latest commit.

maxgolov commented on api/include/opentelemetry/logs/severity.h:
```diff
@@ -0,0 +1,57 @@
+enum class Severity : uint8_t
+{
+  kInvalid = 0,
+  kTrace   = 1,
```
This looks cool and quite readable. But `If the first enumerator does not have an initializer, the associated value is zero. For any other enumerator whose definition does not have an initializer, the associated value is the value of the previous enumerator plus one.` ... basically you do not have to write `= 0`, `= 1`, etc.. Is there a coding style standard that forces us to write more? Since we can safely write less.

xukaren replied:
That makes sense, here since we don't have any non-continuous severity number mappings and the first one starts at 0, we could simplify by removing all the initializers. Will change this in the next commit!

maxgolov commented on sdk/include/opentelemetry/sdk/logs/attribute_utils.h:
```diff
@@ -0,0 +1,123 @@
+ * replaces all non-owning references with owned copies.
+ */
+using LogRecordAttributeValue = nostd::variant<bool,
+                                               int32_t,
```
❤️ the fact that you use standard types here, i.e. `int32_t` - looks much cleaner than `int` we had in another class for traces. Good job!

xukaren replied:
Can't take credit for this - these variants are already defined in `AttributeValue` :)

maxgolov commented on sdk/include/opentelemetry/sdk/logs/attribute_utils.h:
```diff
@@ -0,0 +1,123 @@
+  template <typename T, typename U = T>
+  LogRecordAttributeValue convertSpan(nostd::span<const U> vals)
+  {
+    std::vector<T> copy;
+    for (auto &val : vals)
+    {
+      copy.push_back(T(val));
+    }
```
```suggestion
    const std::vector<T> copy(vals.begin(), vals.end());
```

maxgolov commented on sdk/src/logs/logger.cc:
```diff
@@ -36,23 +51,76 @@ void Logger::log(const opentelemetry::logs::LogRecord &record) noexcept
-   */
-  auto record_pointer =
-      std::unique_ptr<opentelemetry::logs::LogRecord>(new opentelemetry::logs::LogRecord(record));
+  auto recordable = processor->MakeRecordable();
```
I do realize that we allow the SDK built with or without exceptions. However, if we build it with exceptions, is there a provision for the `processor->MakeRecordable()` to throw `std::bad_alloc`? What happens here, is in environment where exceptions are enabled - we'd be silently dropping telemetry. This may not be good in some cases, i.e. if it's a security audit log - we need to potentially bubble up the error condition to the caller, for the caller to decide what to do with that.

I do not think that this is blocking. But if you can log an issue about this, that'd work fine.

maxgolov replied:
Okay, I see that this should be handled by new in `MakeRecordable()` method potentially throwing. Please ignore my above comment.

xukaren replied:
If recordable is null, maybe an error could be invoked using #408?

xukaren replied:
If the processor returns a `nullptr`, could we note an error like:
 ```
if (recordable == nullptr) { 
     error "recordable creation failed";
     return;
}
```` 
where the error is handled by error diagnostics in #408? We may not know the reason for it if the processor/exporter handled it, but at least the user will know something went wrong

ThomsonTan replied:
This could be a workaround. Could you add an `TODO` in the `if` check to add such error diagnostic?

xukaren replied:
Sure, added `TODO` and just opened up https://github.com/open-telemetry/opentelemetry-cpp/issues/441 for further discussion

maxgolov commented on sdk/src/logs/logger.cc:
```diff
@@ -36,23 +51,76 @@ void Logger::log(const opentelemetry::logs::LogRecord &record) noexcept
-   * TODO: Change the API log(LogRecord) function to log(*LogRecord) so the following line
-   * converting record a heap variable can be removed
-   */
-  auto record_pointer =
```
So this code was throwing `std::bad_alloc` before your change, right?

xukaren replied:
Not that I'm aware of, but updated this to use a recordable for faster/more reliable perfomance  (#403), and also allow custom recordable implementations by exporters

maxgolov commented on sdk/src/logs/logger.cc:
```diff
@@ -36,23 +51,76 @@ void Logger::log(const opentelemetry::logs::LogRecord &record) noexcept
+  }
+
+  // Populate recordable fields
+  if (timestamp == this->default_timestamp_)
```
Related to my other comment, that inline timestamp when we logged - would be different from when we populate it here, since this assignment actually has to wait for the lock on `logger_provider_.lock()` and a memory allocation. In case of a congested system it might be a few milliseconds difference.

xukaren replied:
That's a good point, I hadn't thought of that. Would getting the actual timestamp as early as the API be a time-consuming task though? If the LogRecord were filtered out by the Sampler in the SDK for example, would the previous timestamp call have been wasted/unecessary?

maxgolov replied:
It depends. In some scenarios you'd rather prefer the log time as close to original as possible, as you need the actual time when something happened - not when you got thru the lock to assign some timestamp in the SDK, esp. given it's under some lock. This is not a blocker for your PR. I'm thinking maybe we should have a method added in API later on that allows to customize this behavior: whether it's the API stamps the `now` time _right away_ , or if API stamps the indicator `please stamp the time for me` (like you are doing it here), and then later on the next stage stamps it.

maxgolov reviewed (approved):
Good, with minor suggestions.

lalitb commented on api/include/opentelemetry/logs/logger.h:
```diff
@@ -43,81 +60,105 @@ class Logger
+                         resource.begin(), resource.end()},
+                     nostd::span<const std::pair<nostd::string_view, common::AttributeValue>>{
+                         attributes.begin(), attributes.end()},
+                     trace_id, span_id, trace_flags);
```
Nice to see this `initializer_list` for `resource` and `attributes`.

lalitb commented on sdk/include/opentelemetry/sdk/logs/attribute_utils.h:
```diff
@@ -0,0 +1,118 @@
+};
+}  // namespace logs
+}  // namespace sdk
+OPENTELEMETRY_END_NAMESPACE
```
Is it possible we can share this class across trace and logs, and move it to `sdk/common`. Probably not as part of this PR if it's huge task, we can have a issue open for it.

xukaren replied:
I can fix this in a future PR! (If I don't have time I will create an issue for it)

xukaren replied:
@lalitb @jsuereth Just an update on this-  I've moved the `attribute_utils.h` file from `logs` to `common` for this PR, and renamed `LogRecordAttributeValue` to `OwnedAttributeValue`. All the necessary naming changes for `LogRecordAttributeValue` have been done in this PR. 

Seeing as this PR seems to already be running on the long side, and also unrelated to traces/spans, I've filed the traces area needing to be refactored as an issue here: #442 , which I may take a look at when working on the OStream exporter for logs, or if not would probably be a good first issue for someone to tackle as well

lalitb reviewed (approved):
Thanks for working on this PR. We are now using the right approach to propagate logs to exporter : )

ThomsonTan commented on api/include/opentelemetry/logs/logger.h:
```diff
@@ -35,6 +41,17 @@ namespace logs
+  nostd::string_view default_name_;          // default 0 length, nullptr data
+  nostd::string_view default_body_;  // default 0 length, nullptr data. TODO: currently a simple
+                                     // string, but should be changed "Any" type
+  trace::TraceId default_trace_id_;  // default 00000000000000000000000000000000
```
Are these numbers in in hex? Better to have prefix like `0x` to make the indication clear.

xukaren replied:
Yeah `0x00000000000000000000000000000000` would be the value of the TraceId if it were converted to hex. But internally Do we refer to TraceIds or SpanIds in Hex? (i.e. or just simple as 

Internally, it is a 32 `uint8_t` array all initialized to `0`s, but it would be converted to lowercase base 16 representation before being used. 

Would it be simpler to say "default value all initialized to 0s"?

xukaren replied:
Does this follow the convention for Trace/SpanIds? Or would it simply be better to say `default 00000000000000000000000000000000, in base 16`

ThomsonTan replied:
Yes, the suggestion looks good too.

ThomsonTan commented on api/include/opentelemetry/logs/logger.h:
```diff
@@ -43,81 +60,105 @@ class Logger
+  /** Wrapper methods that the user could call for convenience when logging **/
+
+  // Set default values for unspecified fields, then call the base Log() method
+  inline void Log(Severity severity,
```
The `inline` keywords in this header file seems unnecessary because implicit inline applied here?

https://stackoverflow.com/questions/28183729/why-i-can-implement-the-classs-member-function-in-header-file

ThomsonTan commented on api/include/opentelemetry/logs/logger.h:
```diff
@@ -43,81 +60,105 @@ class Logger
-    r.timestamp = time;
-
-    log(r);
+    this->Log(time, severity, default_name_, message, {}, {}, default_trace_id_, default_span_id_,
```
Could the `nullptr` be passed into `Log` instead of `{}` which may involves object construction?

xukaren replied:
Looked into this. I believe since the `KeyValueIterable` is a `const &` reference, and not a pointer itself, it cannot point to a `nullptr` and needs to point to an object with allocated memory (https://stackoverflow.com/questions/4364536/is-null-reference-possible). In this case I believe `{}` should be the best option here

ThomsonTan commented on sdk/include/opentelemetry/sdk/logs/attribute_utils.h:
```diff
@@ -0,0 +1,118 @@
+  LogRecordAttributeValue operator()(bool v) { return LogRecordAttributeValue(v); }
+  LogRecordAttributeValue operator()(int32_t v) { return LogRecordAttributeValue(v); }
+  LogRecordAttributeValue operator()(uint32_t v) { return LogRecordAttributeValue(v); }
+  /*LogRecordAttributeValue operator()(int v)
```
Could this comment about `operator()(int v)` just be removed?

ThomsonTan commented on sdk/src/logs/logger.cc:
```diff
@@ -36,23 +51,76 @@ void Logger::log(const opentelemetry::logs::LogRecord &record) noexcept
+
+  // Inject trace_id/span_id/trace_flags if none is set by user
+  auto provider     = opentelemetry::trace::Provider::GetTracerProvider();
+  auto tracer       = provider->GetTracer("foo_library");
```
What is "foo_library", default tracer name?

xukaren replied:
I didn't notice I left this in - from an earlier example written. Yes, it seems `GetTracer()` doesn't accept an empty param list? Would it be better to remove injection from this PR completely, as it seems to be a separate topic? Or is there a better default name I could use here (which I could also revisit the naming in a later PR)?

xukaren replied:
Fixed - added Logger name and updated the Tracer name above to be the same in latest commit

xukaren commented:
@pyohannes would you be able to review this? @jsuereth I think this addresses some of the concerns you had about the performance and efficiency concerns of the previous approach for log exporting

pyohannes commented on api/include/opentelemetry/logs/logger.h:
```diff
@@ -35,89 +41,118 @@ namespace logs
 
-    log(r);
+  // Set default values for unspecified fields, then call the base Log() method
+  void Log(Severity severity, nostd::string_view message, core::SystemTimestamp time) noexcept
```
Why do we have `time` first in the virtual `Log` call, but `severity` first here? I think we should keep the order of the arguments consistent.

xukaren replied:
I think it would be most ergonomic to users to specify severity, then message. The timestamp overload at the end I think could be an optional argument (which I can now add if you like that idea)

xukaren replied:
It seems strange to specify the timestamp first since it won't be a commonly used feature. Hmm since it seems most ergonomic to users to specify severity, then message, I can move timestamp to the end of the main Log() method for consistency here. Would that work?

pyohannes commented on api/include/opentelemetry/logs/logger.h:
```diff
@@ -35,89 +41,118 @@ namespace logs
 class Logger
 {
+protected:
+  // Default values to set for fields that the user doesn't specify
+  Severity default_severity_;        // default 0 (severity of kInvalid)
+  nostd::string_view default_name_;  // default 0 length, nullptr data
+  nostd::string_view default_body_;  // default 0 length, nullptr data. TODO: currently a simple
+                                     // string, but should be changed "Any" type
+  trace::TraceId default_trace_id_;  // default 00000000000000000000000000000000, in base 16
+  trace::SpanId default_span_id_;    // default 0000000000000000, in base 16
+  trace::TraceFlags default_trace_flags_;  // default 00, in base 16
```
We should get rid of this. I think we can specify `""` whereever `default_name_` or `default_body_` is used, and we could could use `{}` instead of `default_trace_id_`, `default_span_id_` and `default_trace_flags_`.

xukaren replied:
Sure, that sounds pretty intuitive - didn't know `{}` could be a value for trace/spanid :') 

Would you say it's better a default Severity be left as is (to `kInvalid=0`) or directly to `kInfo` as default? Would there be any benefits of knowing when the user intentionally set the severity to `kInfo` and when it was overwritten by the API overloads / SDK?

pyohannes reviewed (commented):
Thanks for refactoring this! That's much closer to a design that is production-ready.

xukaren commented on sdk/src/logs/logger.cc:
```diff
@@ -34,25 +49,70 @@ void Logger::log(const opentelemetry::logs::LogRecord &record) noexcept
+  // Traceid
+  if (!trace_id.IsValid())
+  {
+    recordable->SetTraceId(span_context.trace_id());
```
Added this check

jsuereth commented on sdk/src/logs/logger.cc:
```diff
@@ -34,25 +50,70 @@ void Logger::log(const opentelemetry::logs::LogRecord &record) noexcept
+  }
+  else
+  {
+    recordable->SetTraceId(trace_id);
```
with the new check above, it's possible you're setting trace_id here if `span_context.trace_id().Isvalid()` returns false, i.e. it could be invalid.

Might be easier just do:

```
if trace_id valid => setTraceId
else if context trace id valid => setTraceId
else // nothing.
```

xukaren replied:
This is much cleaner :) Thanks, updated as above (minus the empty `else` block)

jsuereth commented on sdk/include/opentelemetry/sdk/logs/logger.h:
```diff
@@ -41,18 +40,40 @@ class Logger final : public opentelemetry::logs::Logger
   std::weak_ptr<LoggerProvider> logger_provider_;
+
+  // The name of this logger
+  opentelemetry::nostd::string_view logger_name_;
```
Note: you may have to take ownership of this string. 

You need to at least do one of two things:

1. Document in the constructor that the instantiator of the Logger is responsible for keeping this string accessible for the lifetime of the logger.
2. Copy the String into memory you own and can control the lifecycle.

jsuereth reviewed (commented):
Nice work!  Just a few minor nits from me, but thanks very much for tackling this!!

jsuereth reviewed (approved):
Great work!

xukaren commented:
@pyohannes @lalitb Should this be ready to merge?

xukaren commented:
@pyohannes @lalitb Would this be good to merge?

lalitb commented:
@xukaren  - Please rebase the branch with base, so we should be ready to merge it. thanks.

xukaren commented:
> @xukaren - Please rebase the branch with base, so we should be ready to merge it. thanks.

Done!
